tests/test_integration.py

```diff
@@ -0,0 +1,540 @@
+    result = tmp_run("riot --file rf.py", tmp_path)
+    assert (
+        result.stderr
+        == """
+Usage: riot [OPTIONS] COMMAND [ARGS]...
+Try 'riot --help' for help.
+
+Error: Invalid value for '-f' / '--file': Path 'rf.py' does not exist.
+""".lstrip()
+    )
+    assert result.returncode == 2
+
+    rf_path = tmp_path / "rf"
+    rf_path.write_text(
+        """
+from riot import Venv
+venv = Venv()
+""",
+    )
+    result = tmp_run("riot --file rf list")
+    assert (
+        result.stderr
+        == """
+Failed to construct config file:
+Invalid file format for riotfile. Expected file with .py extension got 'rf'.
+""".lstrip()
+    )
+    assert result.returncode == 1
+
+    rf_path = tmp_path / "riotfile.py"
+    rf_path.write_text(
+        """
+from riot import Venv
+venv = Venv()typo1234
+""",
+    )
+    result = tmp_run("riot --file riotfile.py list")
+    assert (
+        """
+Failed to construct config file:
+Failed to parse riotfile 'riotfile.py'.
+""".lstrip()
+        in result.stderr
+    )
+    assert (
+        """
+SyntaxError: invalid syntax
+""".lstrip()
+        in result.stderr
+    )
+    assert result.returncode == 1
+
+
+def test_help(tmp_path: pathlib.Path, tmp_run: _T_TmpRun) -> None:
+    result = tmp_run("riot --help")
+    assert (
+        result.stdout
+        == """
+Usage: riot [OPTIONS] COMMAND [ARGS]...
+
+Options:
+  -f, --file PATH  [default: riotfile.py]
+  -v, --verbose
+  -d, --debug
+  --version        Show the version and exit.
+  --help           Show this message and exit.
+
+Commands:
+  generate  Generate base virtual environments.
+  list      List all virtual env instances matching a pattern.
+  run       Run virtualenv instances with names matching a pattern.
+""".lstrip()
+    )
+    assert result.stderr == ""
+    assert result.returncode == 0
+
+
+def test_version(tmp_path: pathlib.Path, tmp_run: _T_TmpRun) -> None:
+    result = tmp_run("riot --version")
+    assert result.stdout.startswith("riot, version ")
+    assert result.stderr == ""
+    assert result.returncode == 0
+
+
+def test_list_no_file_empty_file(tmp_path: pathlib.Path, tmp_run: _T_TmpRun) -> None:
+    result = tmp_run("riot list")
+    assert (
+        result.stderr
+        == """
+Usage: riot [OPTIONS] COMMAND [ARGS]...
+Try 'riot --help' for help.
+
+Error: Invalid value for '-f' / '--file': Path 'riotfile.py' does not exist.
+""".lstrip()
+    )
+    assert result.returncode == 2
+
+    rf_path = tmp_path / "riotfile.py"
+    rf_path.write_text(
+        """
+from riot import Venv
+""",
+    )
+    result = tmp_run("riot list")
+    assert result.stderr == ""
+    assert result.stdout == ""
+    assert result.returncode == 0
+
+
+def test_list_configurations(tmp_path: pathlib.Path, tmp_run: _T_TmpRun) -> None:
+    rf_path = tmp_path / "riotfile.py"
+    rf_path.write_text(
+        """
+from riot import Venv
+venv = Venv(
+    name="test",
+    pys=[3],
+    command="echo hi",
+)
+""",
+    )
```

Comment:
This still feels like a lot of boilerplate, but I am ok with it.

Reply:
The only boilerplate-y thing is the `rf_path = tmp_path / "riotfile.py"` which really isn't that bad IMO. The alternative you have below `@with_riotfile("riotfile.py")` is pretty much the same thing, just declarative rather than imperative.

We could even write these as one liners:

```python
(tmp_path / "riotfile.py").write_text("""
""")
```

Reply:
yeah, true.